### PR TITLE
Add RFC for KerasCV Single Stage Object Detection.

### PR DESCRIPTION
**Pull request by tanzhenyu**

### CODE REVIEW

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+# Considering a COCO dataset
+coco_dataset = tfds.load('coco/2017')
+train_ds, eval_ds = coco_dataset['train'], coco_dataset['validation']
+
+def preprocess(features):
+  image, gt_boxes, gt_labels = features['image'], features['objects']['bbox'], features['objects']['label']
+  # preprocess image, gt_boxes, gt_labels, such as flip, resize, and padding, and reserve 0 for background label.
+  return image, gt_boxes, gt_labels
+
+anchor_generator = keras_cv.ops.AnchorGenerator(anchor_sizes, scales, aspect_ratios, strides)
```
Comment (fchollet) [inline rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:55]: What are the different submodules in `keras_cv`? Everything seems to live in `ops`.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+anchor_generator = keras_cv.ops.AnchorGenerator()
+anchor_boxes = anchor_generator(image_size)
+detection_generator = keras_cv.layers.NMSDetectionDecoder()
+@tf.function
+def serving_fn(image):
+  batched_image = tf.expand_dims(image)
+  raw_boxes, scores = loaded_model(batched_image, training=False)
+  decoded_boxes = box_coder.decode(raw_boxes, anchor_boxes)
+  classes, scores, boxes, _ = detection_generator(scores, decoded_boxes)
+  return classes, scores, boxes
```
Comment (fchollet) [inline rfcs/20200928-keras-cv-single-stage-2d-object-detection.md]: Consider returning a dictionary

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 1 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:55]: So we proposed 2 layers, 1 loss, and 4 ops. However I'd like your input on whether `BoxCoder` and `BoxMatcher` should be layer instead, given our example workflow.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 2 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md]: Done.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+coco_dataset = tfds.load('coco/2017')
+train_ds, eval_ds = coco_dataset['train'], coco_dataset['validation']
+
+def preprocess(features):
+  image, gt_boxes, gt_labels = features['image'], features['objects']['bbox'], features['objects']['label']
+  # preprocess image, gt_boxes, gt_labels, such as flip, resize, and padding, and reserve 0 for background label.
+  return image, gt_boxes, gt_labels
+
+anchor_generator = keras_cv.ops.AnchorGenerator(anchor_sizes, scales, aspect_ratios, strides)
+similarity_calculator = keras_cv.ops.IOUSimilarity()
```
Comment (fchollet) [inline rfcs/20200928-keras-cv-single-stage-2d-object-detection.md]: "layers"?

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+class IouSimilarity(tf.keras.layers.Layer):
+  """Class to compute similarity based on Intersection over Union (IOU) metric."""
+ 
+  def __init__(self, mask_value):
+    """Initializes IouSimilarity layer.
+    Args:
+      mask_value: A float mask value to fill where `mask` is True. 
+    """
+ 
+  def __call__(self, groundtruth_boxes, anchors, mask=None):
```
Comment (fchollet) [inline rfcs/20200928-keras-cv-single-stage-2d-object-detection.md]: call

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+
+Single-stage object detection models are a state-of-art technique that powers many computer vision tasks, they provide
+faster detection compared to two-stage models (such as FasterRCNN), while maintaining comparable performance.
+
+With this proposal, Keras users will be able to build end-to-end models with a simple API.
+
+## Design overview
+
+This proposal includes the specific core components for building single-stage object detection models. It does not,
+include, however, include:
```
Comment (rola93) [inline rfcs/20200928-keras-cv-single-stage-2d-object-detection.md]: >  It does not,
include, however, include:

Those lines are confussing

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+  def __init__(self,
+               alpha,
+               gamma,
+               reduction=tf.keras.losses.Reduction.AUTO,
+               name=None):
+    """Initializes `FocalLoss`.
+
+    Arguments:
+      alpha: The `alpha` weight factor for binary class imbalance.
+      gamma: The `gamma` focusing parameter to re-weight loss.
```
Comment (rola93) [inline rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:248]: Should not include default values for `alpha` an `gamma`?

**Review by rola93 — COMMENTED** [review]
Awesome work!

Just small suggestion on confusing line (at least for me).

Also a comment with respect to default values which may apply to other functions: shouldn't them be supplied?

Thanks for your work!

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+    # self.head = classification and regression heads
+  
+  def call(self, image, training=None):
+    feature_map = self.backbone(image, training)
+    feature_map = self.fpn(feature_map, training)
+    class_scores, boxes = self.head(feature_map, training)
+    return {'classification': class_scores, 'regression': boxes}
+
+transformed_train_ds = train_ds.map(preprocess).map(encode_label).batch(128).shuffle(1024)
+transformed_eval_ds = eval_ds.map(preprocess).map(encode_label).batch(128)
```
Comment (haifeng-jin) [inline rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:96]: Should we have a preprocessing layer for the image to make the model more encapsulated?

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+`keras.applications` for heavily used patterns, however the user can build them easily using Keras.
+
+#### Training
+
+Case where a user want to train from scratch:
+
+```python
+import tensorflow as tf
+import tensorflow_datasets as tfds
+import keras_cv
```
Comment (DrChrisLevy) [inline rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:44]: I'm use to using `tf.keras`. What is `keras_cv`?
Can't this stuff live within `tf.keras.applications`?

Also, this seems related to my feature request here: https://github.com/tensorflow/tensorflow/issues/42393#
This is exciting!

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+
+Data augmentation will be included as a separate RFC that handles a
+broader context than object detection.
+
+Model backbone and detection heads are model-specific, we anticipate them to be analyzed and proposed in 
+`keras.applications` for heavily used patterns, however the user can build them easily using Keras.
+
+#### Training
+
+Case where a user want to train from scratch:
```
Comment (DrChrisLevy) [inline rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:39]: Would there be the option to train from a pre-defined checkpoint? For example, start from
a model pre-trained on coco data set and then do transfer learning. Having those weights
from coco for example for the different backbones in `tf.keras.applications` would be nice.
Much like we have the imagenet weights and there is the option to remove the head and have the pre-trained
feature extractor (as it is today for classification for example in `tf.keras.applications`).

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 5 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:24]: Thanks.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 4 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md]: Done.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 3 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md]: Done.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 6 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:248]: Sounds good.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 7 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:96]: Data augmentation is not part of this proposal.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 8 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:44]: You'd probably have to briefly surf through the scoping doc:
https://github.com/keras-team/governance/blob/master/rfcs/20200827-keras-cv-scoping-design.md

Essentially, keras_cv is for meeting complicated modeling demands such as object detections, and segmentations. We will refactor some of keras_application utilities so that keras_cv can rely on the backbone building

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 9 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:39]: Yes! I forgot to mention this. Updated.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 6 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:248]: Sure.. Done.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 5 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md]: Updated.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 4 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:151]: Done.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 3 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:56]: Done.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 8 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:44]: You'd probably have to briefly surf the scoping doc:
https://github.com/keras-team/governance/blob/master/rfcs/20200827-keras-cv-scoping-design.md

Essentially, keras_cv is for meeting more complicated modeling demands, such as detection and segmentation. It will directly rely on keras-application for backbone building blocks.

File: rfcs/20200928-keras-cv-single-stage-2d-object-detection.md (hunk 7 above)
Comment (tanzhenyu) [reply rfcs/20200928-keras-cv-single-stage-2d-object-detection.md:96]: Data augmentation is not part of this proposal. But yes, it is part of the roadmap.
